src/flp/types/dp.rs

```diff
@@ -0,0 +1,179 @@
+use num_integer::Integer;
+use num_traits::{One, Pow};
+use rand::{distributions::Distribution, Rng, SeedableRng};
+
+// TODO(#1071): This is implemented for the concrete fields `Field64` and `Field128` in order to
+// avoid imposing the `BigInt: From<F::Integer>` bound on all callers. In the future, we may want to
+// impose this bound on `FieldElementWithInteger`, which would allow simple blanket implementations
+// here.
+impl<S> TypeWithNoise<PureDpDiscreteLaplace> for SumVec<Field64, S>
+where
+    S: ParallelSumGadget<Field64, Mul<Field64>> + Eq + 'static,
+{
+    fn add_noise_to_result(
+        &self,
+        dp_strategy: &PureDpDiscreteLaplace,
+        agg_result: &mut [Self::Field],
+        _num_measurements: usize,
+    ) -> Result<(), FlpError> {
+        self.add_noise(
+            dp_strategy,
+            agg_result,
+            &mut SeedStreamTurboShake128::from_entropy(),
+        )
+    }
+}
+
+impl<S> TypeWithNoise<PureDpDiscreteLaplace> for SumVec<Field128, S>
+where
+    S: ParallelSumGadget<Field128, Mul<Field128>> + Eq + 'static,
+{
+    fn add_noise_to_result(
+        &self,
+        dp_strategy: &PureDpDiscreteLaplace,
+        agg_result: &mut [Self::Field],
+        _num_measurements: usize,
+    ) -> Result<(), FlpError> {
+        self.add_noise(
+            dp_strategy,
+            agg_result,
+            &mut SeedStreamTurboShake128::from_entropy(),
+        )
+    }
+}
+
+impl<F, S> SumVec<F, S>
+where
+    F: FftFriendlyFieldElement,
+    BigInt: From<F::Integer>,
+    F::Integer: TryFrom<BigInt, Error = TryFromBigIntError<BigInt>>,
+{
+    fn add_noise<R>(
+        &self,
+        dp_strategy: &PureDpDiscreteLaplace,
+        agg_result: &mut [F],
+        rng: &mut R,
+    ) -> Result<(), FlpError>
+    where
+        R: Rng,
+    {
+        // Compute the l1-sensitivity of the aggregation function (assuming the substitution-DP
+        // model). The worst case is when one individual's measurement changes such that each vector
+        // element flips from 0 to 2^bits - 1, or vice versa. Then, the l1 distance from the initial
+        // query result to the new query result will be (2^bits - 1) * length.
+        let two = BigUint::from(2u64);
+        let bits = BigUint::from(self.bits);
+        let length = BigUint::from(self.len);
+        let sensitivity = (Pow::pow(two, &bits) - BigUint::one()) * length;
+
+        // Initialize sampler.
+        let sampler = dp_strategy.create_distribution(sensitivity.into())?;
+
+        // Generate noise for each vector coordinate and apply it.
+        let modulus = BigInt::from(F::modulus());
+        for entry in agg_result.iter_mut() {
+            // Generate noise.
+            let noise = sampler.sample(rng);
+
+            // Project it into the field by taking the modulus, converting to a fixed-precision
+            // integer, and converting to a field element.
+            let noise_wrapped = noise.mod_floor(&modulus);
```

Comment:
Question: does this function throw an exception if for example, `noise` is more than `modulus`?

Another consideration is: because `noise` could be negative, we may need to reserve some tail end of the field modulus for negative values in aggregate result, so there may need to be a reasonable upper bound on the magnitude of `noise`, perhaps `|noise| < modulus/2`?

Reply:
No, this succeeds unconditionally, and would wrap around on very large noise values. The rationale, (see https://github.com/divviup/libprio-rs/pull/578#discussion_r1226543791 and https://github.com/divviup/libprio-rs/pull/578#discussion_r1263075241) is that adding wrapped noise in the field, then unsharding to produce an integer, is equivalent to a trusted curator adding noise to the true sum using arbitrary-precision integers, and then taking the modulus as a post-processing step. (and post-processing preserves DP guarantees) I should bake this reasoning into a comment when I extract this loop into a helper function.

The `mod_floor()` operation does wrap negative noise values around to positive values below `modulus`; we're fine with this overlapping with the extreme upper tail of the noise distribution.

Reply:
I see, makes sense. Out of curiosity, how do you plan to interpret the aggregate result with noise in this case? Given a reasonably large field, and assuming number of clients cannot exceed the field modulus, how much of the tail end of the field modulus do you plan to reserve for noise?

Reply:
I hadn't thought about it too much. I suppose the optimal choice will depend on the batch size (and the VDAF's type and parameters).